core/src/mast/mod.rs

```diff
@@ -133,11 +133,62 @@ impl MastForest {
     }
 
     /// Removes all decorators from this MAST forest.
+    ///
+    /// This method modifies the forest in-place, removing all decorator information
+    /// including operation-indexed decorators, before-enter decorators, after-exit
+    /// decorators, and error codes.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use miden_core::mast::MastForest;
+    ///
+    /// let mut forest = MastForest::new();
+    /// // Add decorators and nodes to the forest
+    /// forest.strip_decorators(); // forest is now stripped
+    /// ```
     pub fn strip_decorators(&mut self) {
         // Clear all debug info (decorators and error codes)
         self.debug_info.clear();
     }
 
+    /// Compacts the forest by merging duplicate nodes.
+    ///
+    /// This operation performs node deduplication by merging the forest with itself.
+    /// The method assumes that decorators have already been stripped if that is desired.
+    /// The operation modifies the forest in-place, updating all node references as needed.
+    ///
+    /// The process works by:
+    /// 1. Merging the forest with itself to deduplicate identical nodes
+    /// 2. Updating internal node references and remappings
+    /// 3. Modifying the forest in-place with the compacted result
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use miden_core::mast::MastForest;
+    ///
+    /// let mut forest = MastForest::new();
+    /// // Add nodes to the forest
+    ///
+    /// // First strip decorators if needed
+    /// forest.strip_decorators();
+    ///
+    /// // Then compact the forest
+    /// forest.compact().unwrap();
+    ///
+    /// // Forest is now compacted with duplicate nodes merged
+    /// ```
+    pub fn compact(&mut self) -> Result<MastForestRootMap, MastForestError> {
+        // Merge with itself to deduplicate nodes
+        let (compacted_forest, root_map) = MastForest::merge([&*self])?;
```

Comment:
Under what conditions can merging a forest with itself fail? I'm wondering if we can replace this with `expect()` and avoid returning the result.